Refactor the E2E test infra

chizhg opened the pull request:
Fixes #

## Proposed Changes

-
-
-

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
NONE
```

/hold
/cc

chizhg commented:
/test pull-google-knative-gcp-unit-tests

chizhg commented:
/test pull-google-knative-gcp-build-tests

chizhg commented:
/test pull-google-knative-gcp-build-tests

chizhg commented:
/test pull-google-knative-gcp-integration-tests

chizhg commented:
/test pull-google-knative-gcp-integration-tests

chizhg commented:
/retest

chizhg commented:
/retest

chizhg commented:
We haven't added the e2e test for Scheduler, but now the PR is ready for the initial review.

/cc @nachocano 
/cc @capri-xiyue

chizhg commented:
`TestAuditLogsSource` is still flaky after I changed the `TIME_OUT` to 360, FYI @capri-xiyue 
/test pull-google-knative-gcp-integration-tests

nachocano commented on test/e2e-common.sh:
```diff
@@ -151,9 +169,24 @@ function teardown() {
   fi
   kubectl -n ${E2E_TEST_NAMESPACE} delete secret ${PUBSUB_SECRET_NAME}
-  # TODO delete control plane secret?
+  kubectl -n ${CONTROL_PLANE_NAMESPACE} delete secret ${CONTROL_PLANE_SECRET_NAME}
```
if you delete the secret, but there are some resources in that namespace, say a topic, pullsubscription, or whatever, then they won't be able to be deleted... are we cleaning up those before right?

chizhg replied:
Ah, we missed `knative_teardown` in this script..
I'll add it later in this PR.

chizhg replied:
I made some changes to only delete the Google cloud service account and iam bindings in the teardown phase, as other resources will be deleted automatically when the cluster is deleted.

nachocano commented on test/e2e/lib/creation.go:
```diff
@@ -0,0 +1,117 @@
+func (c *Client) CreateUnstructuredObjOrFail(spec *unstructured.Unstructured) {
+	namespace := c.Namespace
+	gvr, _ := meta.UnsafeGuessKindToResource(spec.GroupVersionKind())
+	_, err := c.Core.Dynamic.Resource(gvr).Namespace(namespace).Create(spec, v1.CreateOptions{})
```
nit: just use c.Namespace and don't need for that auxiliary namespace variable? Same as below

chizhg replied:
Done.

nachocano commented on test/e2e/lib/creation.go:
```diff
@@ -0,0 +1,117 @@
+		namespace := job.Namespace
+		svc := resources.ServiceDefaultHTTP(name, job.Spec.Template.Labels)
+
+		svcs := client.Core.Kube.Kube.CoreV1().Services(namespace)
```
similar as with previous comment, use `job.Namespace`? No strong opinion, but I'd rather minimize creation of temp variables if possible.

chizhg replied:
Done.

nachocano commented on test/e2e/lib/creation.go:
```diff
@@ -0,0 +1,117 @@
+func (c *Client) CreateJobOrFail(job *batchv1.Job, options ...func(*batchv1.Job, *Client) error) {
+	// set namespace for the job in case it's empty
+	namespace := c.Namespace
+	job.Namespace = namespace
```
but if it's set, you are overriding it? Maybe:

`if job.Namespace == "" {
    job.Namespace = c.Namespace
}`

chizhg replied:
Done. Thanks.

nachocano commented on test/e2e/lib/lifecycle.go:
```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2019 Google LLC
```
2020

nachocano commented on test/e2e/lib/lifecycle.go:
```diff
@@ -0,0 +1,243 @@
+	return operations.GetFirstTerminationMessage(pod), nil
+}
+
+func (c *Client) LogsFor(namespace, name string, tm *metav1.TypeMeta) (string, error) {
```
I've seen this somewhere else, worth moving to knative/pkg/test as well? maybe add a TODO as with the previous one?

chizhg replied:
Added a TODO comment.

nachocano commented on test/e2e/lib/resources/kube.go:
```diff
@@ -0,0 +1,71 @@
+	return baseJob(name, "sender", envVars)
+}
+
+func baseJob(name, imageName string, envVars []v1.EnvVar) *batchv1.Job {
```
maybe add a comment as all of the above use them?

chizhg replied:
Added the comment.

nachocano commented on test/e2e/lib/resources/unstructured.go:
```diff
@@ -0,0 +1,68 @@
+	pkgTest "knative.dev/pkg/test"
+)
+
+func IstioServiceEntry(name, namespace string) *unstructured.Unstructured {
```
comment?

chizhg replied:
Done.

nachocano commented on test/e2e/lib/resources/unstructured.go:
```diff
@@ -0,0 +1,68 @@
+	return &unstructured.Unstructured{Object: obj}
+}
+
+func ReceiverKService(name, namespace string) *unstructured.Unstructured {
```
same here?

chizhg replied:
Done.

nachocano commented on test/e2e/lib/topics.go:
```diff
@@ -52,11 +52,11 @@ func makeTopicOrDie(t *testing.T) (string, func()) {
 }
 
-func makeTopicOrDieWithTopicName(t *testing.T, topicName string) (string, func()) {
+func MakeTopicOrDieWithTopicName(t *testing.T, topicName string) (string, func()) {
```
`MakeTopicWithNameOrDie` ?

chizhg replied:
Changed to the name as suggested.

nachocano commented on test/e2e/lib/typemetas.go:
```diff
@@ -0,0 +1,68 @@
+	}
+}
+
+var ChannelTypeMeta = gcloudMessagingTypeMeta("Channel")
```
why do we need `gcloud` prefix here, and all of the ones below... Is this coupled with gcloud?

chizhg replied:
Removed `gcloud` prefix.

nachocano commented on test/e2e/test_auditlogs.go:
```diff
@@ -34,73 +39,68 @@ import (
+		kngcptesting.WithAuditLogsSourceResourceName(resourceName),
+		kngcptesting.WithAuditLogsSourceSink(metav1.GroupVersionKind{
+			Version: "v1",
+			Kind:    "Service"}, targetName))
```
might be cleaner to save the v1 Service GVK in a variable, as I guess it will be used in many places?

chizhg replied:
Added it as a variable.

nachocano commented on test/e2e/test_broker_pubsub.go:
```diff
@@ -46,119 +49,102 @@ PubSubWithBrokerTestImpl tests the following scenario:
+		dummyTriggerName,
+		eventingtestresources.WithBroker(brokerName),
+		eventingtestresources.WithAttributesTriggerFilter(
+			"", "",
```
what does "", "" refers to? Can you add `/* <description> */` for non-IDE reference...

chizhg replied:
Changed to `eventingv1alpha1.TriggerAnyFilter`

nachocano commented on test/e2e/test_broker_pubsub.go:
```diff
@@ -46,119 +49,102 @@ PubSubWithBrokerTestImpl tests the following scenario:
+		respTriggerName,
+		eventingtestresources.WithBroker(brokerName),
+		eventingtestresources.WithAttributesTriggerFilter(
+			"", "",
```
same as above

chizhg replied:
Ditto.

nachocano commented on test/e2e/test_pubsub.go:
```diff
@@ -24,120 +24,87 @@ import (
+	// Create the PubSub source.
+	eventsPubsub := kngcptesting.NewPubSub(psName, client.Namespace,
+		kngcptesting.WithPubSubSink(metav1.GroupVersionKind{
+			Version: "v1",
```
you can use the saved svcGVR from the previous comment.

nachocano commented on test/e2e/test_pubsub.go:
```diff
@@ -24,120 +24,87 @@ import (
+	client.CreatePubSubOrFail(eventsPubsub)
+
+	// Create the Istio ServiceEntry.
+	istioServiceEntry := resources.IstioServiceEntry(
```
wondering if we are doing this istio thing in eventing as well?
can you remind me why we need it in code, and cannot be done while setting up the cluster in the bash script?

chizhg replied:
No we are not using Istio in eventing test.

I "translated" it directly from the original e2e tests, but after I removed it, the tests were still able to pass except `TestBrokerWithPubSubChannel` (I guess it's because that test is using Serving, which depends on Istio?)

Do you know if the `ServiceEntry` has to be created in the namespace where the test is run? If it does not have to be in the same namespace, we can definitely move it to the bash script.

nachocano replied:
I think it doesn't have to be created in the namespace that the test runs. The only step we do in non-test clusters is to setup the cluster local gateway and things work... It might be doable in the bash script

chizhg replied:
Removed all the relevant code in Go, and created it from the bash script.
Thanks!

nachocano commented on test/e2e/test_pubsub.go:
```diff
@@ -24,120 +24,87 @@ import (
 	}
 
-	topic := getTopic(t, topicName)
+	time.Sleep(1 * time.Minute)
```
1 min? Seems a lot? Maybe add a comment saying why...

chizhg replied:
This is ported from the original test. But I tried to run the test without it, the test can still pass.
So removed it.

nachocano commented on test/e2e/test_pull_subscription.go:
```diff
@@ -20,114 +20,82 @@ import (
+			Topic: topic,
+		}),
+		kngcptesting.WithPullSubscriptionSink(metav1.GroupVersionKind{
+			Version: "v1",
```
svcGVR, just comment it so that you don't miss it.

nachocano commented on test/e2e/test_pull_subscription.go:
```diff
@@ -20,114 +20,82 @@ import (
+	pullsubscription := kngcptesting.NewPullSubscription(psName, client.Namespace,
+		kngcptesting.WithPullSubscriptionSpec(v1alpha1.PullSubscriptionSpec{
+			Topic: topicName,
+		}), kngcptesting.WithPullSubscriptionSink(metav1.GroupVersionKind{
```
svcGVR, just comment it so that you don't miss it.

nachocano commented on test/e2e/test_storage.go:
```diff
@@ -83,64 +87,41 @@ func getBucketHandle(ctx context.Context, t *testing.T, bucketName string, proje
+	// Create the Storage source.
+	eventsStorage := kngcptesting.NewStorage(storageName, client.Namespace,
+		kngcptesting.WithStorageBucket(bucketName),
+		kngcptesting.WithStorageSink(metav1.GroupVersionKind{
```
svcGVR, just comment it so that you don't miss it.

nachocano commented on test/lib.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2019 The Knative Authors
+# Copyright 2020 Google LLC
```
if this was a previous file, I think it should keep the 2019, but change to Google LLC... but not entirely sure.

chizhg replied:
Changed it back to `2019`.

nachocano commented on test/presubmit-tests.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2019 The Knative Authors
+# Copyright 2020 Google LLC
```
same as above

chizhg replied:
Ditto

nachocano commented on vendor/knative.dev/eventing/test/test_images/logevents/main.go:
```diff
@@ -0,0 +1,49 @@
+/*
```
would it make sense to move this test images to pkg? maybe not because is too eventing specific?

chizhg replied:
Yeah it is too eventing specific. Currently all the code under `pkg/test` are generic to all Knative projects.

nachocano commented:
Thanks for doing this @chizhg !!!
Just a few optional nits!

chizhg commented:
> Thanks for doing this @chizhg !!!
> Just a few optional nits!

Thanks for the review @nachocano !
I think I've addressed most of the issues, now this PR is ready for the second round of review.

nachocano commented on test/e2e/test_broker_pubsub.go:
```diff
@@ -93,9 +93,9 @@ func BrokerWithPubSubChannelTestImpl(t *testing.T) {
-	istioServiceEntry := resources.IstioServiceEntry(
-		"cloud-run-events-googleapis-ext", client.Namespace)
-	client.CreateUnstructuredObjOrFail(istioServiceEntry)
+	// istioServiceEntry := resources.IstioServiceEntry(
```
remove this commented lines?

chizhg replied:
Removed.

nachocano commented on test/e2e/test_auditlogs.go:
```diff
@@ -89,11 +89,6 @@ func AuditLogsSourceWithTestImpl(t *testing.T) {
 	client.CreateAuditLogsOrFail(eventsAuditLogs)
 
-	// Create the Istio ServiceEntry.
-	istioServiceEntry := resources.IstioServiceEntry(
```
if this is not used, then you can remove the resources.IstioServiceEntry code?

chizhg replied:
Removed.

nachocano commented on test/e2e/lib/creation.go:
```diff
@@ -113,5 +112,5 @@ func (c *Client) CreateJobOrFail(job *batchv1.Job, options ...func(*batchv1.Job,
 		c.T.Fatalf("Failed to create job %q: %v", job.Name, err)
 	}
-	c.Tracker.Add("batch", "v1", "jobs", namespace, job.Name)
+	c.Tracker.Add("batch", "v1", "jobs", c.Namespace, job.Name)
```
should be job.Namespace instead of c.Namespace?

chizhg replied:
Fixed.

nachocano commented on test/e2e/lib/resources/unstructured.go:
```diff
@@ -21,6 +21,7 @@ import (
 )
 
+// IstioServiceEntry creates a ServiceEntry to add a few Google services to the service mesh.
 func IstioServiceEntry(name, namespace string) *unstructured.Unstructured {
```
this might go away if not used?

chizhg replied:
Removed.

nachocano commented on test/e2e/test_broker_pubsub.go:
```diff
@@ -65,12 +66,22 @@ func BrokerWithPubSubChannelTestImpl(t *testing.T) {
 	client.Core.CreateBrokerOrFail(brokerName, lib.ChannelTypeMeta)
 
+	// Create the Istio ServiceEntry.
+	istioServiceEntry := resources.IstioServiceEntry(
```
Ohh I see you are using it here still

chizhg replied:
You are right, we don't have to create it per namespace. 
I installed it from the Shell script and removed the namespace creation, and it still works :)

nachocano commented on test/e2e/README.md:
```diff
@@ -94,6 +57,54 @@ the `Monitoring Editor` role on your Google Cloud project:
+
+- [`KO_DOCKER_REPO`](https://github.com/knative/serving/blob/master/DEVELOPMENT.md#environment-setup)
+  to be set
+- You to be
```
You need to be?

chizhg replied:
Fixed.

nachocano commented:
/lgtm
/approve

chizhg commented:
/hold cancel
